File: Data Structure/linked_list/double_ll.py
class Node: 
    def __init__(self, data=None, next=None, prev=None):
        self.data = data
        self.next = next
        self.prev = prev

class LinkedList:
    def __init__(self):
        self.head = None

    def get_length(self): 
        count = 0
        curr_node = self.head 
        while curr_node:
            count += 1
            curr_node = curr_node.next 
        return count

    def insert_at_beginning(self, data):
        if self.head == None:
            self.head = Node(data, None, None)
        else:
            node = Node(data, self.head, None)
            self.head.prev = node
            self.head = node

    def insert_at_end(self, data):
        if self.head is None:
            self.head = Node(data, None, None)
            return

        curr_node = self.head 
        while curr_node.next:
            curr_node = curr_node.next

        curr_node.next = Node(data, None, curr_node)
    
    def insert_at(self, index, data): 
        if index < 0 or index >= self.get_length(): 
            raise Exception("Invalid index") 
        if index == 0: 
            self.insert_at_beginning(data)
            return 
    
        count = 0 
        curr_node = self.head

        while (curr_node): 
            if (count == index - 1): 
                new_node = Node(data, curr_node.next, curr_node)
                if curr_node.next:
                    curr_node.next.prev = new_node
                curr_node.next = new_node
                return 
            curr_node = curr_node.next
            count += 1 

    def insert_after_value(self, value, data_to_insert):
        if self.head is None:
            return

        curr_node = self.head 
        while(curr_node):
            if (curr_node.data == value): 
                new_node = Node(data_to_insert, curr_node.next, curr_node)    
                if curr_node.next: 
                    curr_node.next.prev = new_node
                curr_node.next = new_node
                return 

            curr_node = curr_node.next

    # Create a new linked list for the given list 
    def insert_values(self, data_list): 
        self.head = None
        for data in data_list:
            self.insert_at_end(data)
    
    def remove_by_value(self, data):
        if self.head is None:
            return
            
        if (self.head.data == data): 
            self.head = self.head.next
            self.head.prev = None 
            return 

        curr_node = self.head 
        while(curr_node.next):
            if (curr_node.next.data == data): 
                if (curr_node.next.next):  
                    curr_node.next.next.prev = curr_node
                curr_node.next = curr_node.next.next    
                return 

            curr_node = curr_node.next

    def remove_at(self, index): 
        if index < 0 or index >= self.get_length(): 
            raise Exception("Invalid index") 
        if index == 0: 
            self.head = self.head.next
            self.head.prev = None
            return 

        count = 0 
        curr_node = self.head 
        while curr_node:
            if (count == index - 1): 
                if (curr_node.next.next):  
                    curr_node.next.next.prev = curr_node
                curr_node.next = curr_node.next.next    
                return 

            curr_node = curr_node.next
            count += 1 

    def print(self): 
        if (self.head is None): 
            print('Linked list is empty') 
            return
        
        curr_node = self.head 
        linkedlist_str = ''
        while curr_node: 
            linkedlist_str += str(curr_node.data) + ' <---> ' 
            curr_node = curr_node.next
        linkedlist_str += 'None'
        print(linkedlist_str)
    
    def get_last_node(self):
        itr = self.head
        while itr.next:
            itr = itr.next

        return itr

    def print_forward(self):
        if self.head is None:
            print("Linked list is empty")
            return

        curr_node = self.head
        linkedlist_str = ''
        while curr_node:
            linkedlist_str += str(curr_node.data) + ' --> '
            curr_node = curr_node.next
        print(linkedlist_str)

    def print_backward(self):
        if self.head is None:
            print("Linked list is empty")
            return

        last_node = self.get_last_node()
        itr = last_node
        llstr = ''
        while itr:
            llstr += itr.data + ' --> '
            itr = itr.prev
        print("Link list in reverse: ", llstr)

if __name__ == '__main__':
    # ll = LinkedList();
    # ll.insert_at_beginning(5)
    # ll.insert_at_beginning(89)
    # ll.insert_at_end(99)
    # ll.insert_at_end(59)
    # ll.print()
    
    ll2 = LinkedList() 
    ll2.insert_values(['banana', 'orange', 'coconut', 'blueberry', 'mango'])
    ll2.print()
    ll2.insert_at(3, 'watermelon')
    ll2.print()
    ll2.insert_after_value('banana', 'papaya')
    ll2.print()
    ll2.remove_by_value('mango')
    ll2.print()
    ll2.remove_at(2)
    ll2.print()


    # ll3 = LinkedList() 
    # ll3.print()
   

File: Data Structure/linked_list/double_ll.js
"use strict";

class Node {
  constructor(data = null, next = null, prev = null) {
    this.data = data;
    this.next = next;
    this.prev = prev;
  }
}

/***
  LinkedList Operations:
      1. print(): access each element of the linked list and display them
      2. insertAtBeginning(value): add a new node to the beginning 
      2. insertAtEnd(value): add a new node to the end 
      3. insertAt(value, index): add a new node at a specified index   
      5. removeByValue(value): remove a node 
      6. removeAt(index): remove a node at a specified index   
      7. find(value): find a node 
  */

class DoubleLinkedList {
  constructor() {
    this.tail = null;
    this.head = null;
    this.length = 0;
  }

  /**
   * Print the linked list in a concise and clean string representation
   * @param {any types} value
   */
  print() {
    let curr = this.head;
    if (!this.head) {
      console.log("Empty LinkedList");
      return;
    }

    let result = `Head: ${this.head.data}, Tail: ${this.tail.data}, length: ${this.length} \n${curr.data}`;
    while (curr.next) {
      curr = curr.next;
      result += ` <--> ${curr.data}`;
    }
    console.log(result);
  }

  /**
   * Insert a new value to the beginning of the linked list
   * @param {any types} value
   */
  insertAtBeginning(value) {
    this.length += 1;
    const newNode = new Node(value, this.head, null);

    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
      return;
    }

    this.head.prev = newNode;
    this.head = newNode;
  }

  /**
   * Insert a new value to the end of the linked list
   * @param {any types} value
   */
  insertAtEnd(value) {
    this.length += 1;
    const newNode = new Node(value, null, null);

    // Handle empty linked list
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
      return;
    }

    // Handle regular case
    this.tail.next = newNode;
    newNode.prev = this.tail;
    this.tail = newNode;
  }

  /**
   * Insert a new value at a specified index to the linked list
   * @param {any types} value
   * @param {Number} index
   * Improvement (TODO): only search half the list
   */
  insertAt(value, index) {
    if (index > this.length || index < 0)
      throw new Error("Invalid index (out of bound");

    const newNode = new Node(value, null, null);

    // Handle removing node at index 0
    if (index === 0) {
      this.insertAtBeginning(value);
      return;
    }

    if (index === this.length) {
      this.insertAtEnd(value);
      return;
    }

    this.length++;
    const midPoint = Math.floor(this.length / 2);
    let curr;

    if (index <= midPoint) {
      curr = this.head;

      // We want to stop at the position right before the desired index
      for (let i = 0; i < index - 1; i++) {
        curr = curr.next;
      }
    } else {
      curr = this.tail;

      for (let i = this.length; i > index + 1; i--) {
        curr = curr.prev;
      }
    }

    newNode.next = curr.next;
    newNode.prev = curr;
    curr.next.prev = newNode;
    curr.next = newNode;
  }

  /**
   * Remove a node with the given value from the linked list
   * @param {any types} value
   * @returns the linked list length if successfully removed the node, -1 otherwise.
   */
  removeByValue(value) {
    let curr = this.head;

    if (curr === null) return -1;

    if (curr.data === value) {
      this.head = curr.next;
      this.head.prev = null;

      this.length += 1;
      return this.length;
    }

    // Need to stop at the prev node
    while (curr.next) {
      if (curr.next.data === value) {
        // Update this.tail value
        if (curr.next === this.tail) {
          this.tail = curr;
        }

        // remove operation
        curr.next = curr.next.next;
        if (curr?.next?.next?.prev) {
          curr.next.next.prev = curr;
        }
        this.length -= 1;
        return this.length;
      }

      curr = curr.next;
    }
    return -1;
  }

  /**
   * Remove a node at the given index from the linked list
   * @param {Number} index
   * @returns the current if successfully removed the node, -1 otherwise.
   */
  removeAt(index) {
    let curr = this.head;

    // Check for valid index
    if (index < 0 || index > this.length)
      throw new Error("Invalid index (out of bound");

    // return -1 if linked list is empty
    if (curr === null) return -1;

    // Handle remove the 1st node
    if (index === 0) {
      this.head = curr.next;
      curr.next.prev = null;
      this.length -= 1;
      return this.length;
    }

    // Handle remove the last node
    if (index === this.length) {
      this.tail = this.tail.prev;
      this.tail.next = null;
      this.length -= 1;
      return this.length;
    }

    // Handle regular case (we could improve this by only check half the linked list)
    for (let i = 0; i < index - 1; i++) {
      curr = curr.next; // the node before the one we want to remove
    }

    curr.next = curr.next.next ? curr.next.next : null;
    if (curr?.next?.next?.prev) {
      curr.next.next.prev = curr;
    }

    this.length -= 1;
    return this.length;
  }

  /**
   * Find the provided data in the linked list
   * @param {any types} value
   * @returns an index if found, -1 otherwise.
   */
  find(value) {
    let curr = this.head;

    // Handle empty case
    if (!curr) return -1;
    let count = 0;

    while (curr.next) {
      if (curr.data === value) {
        return count;
      }
      count += 1;
      curr = curr.next;
    }
    return -1;
  }
}

/*------- TESTING -----------*/
// const myDoubleLL = new DoubleLinkedList();
// myDoubleLL.insertAtBeginning(1);
// myDoubleLL.insertAtBeginning(2);
// myDoubleLL.insertAtBeginning(3);
// myDoubleLL.print();
// myDoubleLL.insertAtEnd("end");
// myDoubleLL.print();
// myDoubleLL.insertAt("insertAt", myDoubleLL.length);
// myDoubleLL.print();
// myDoubleLL.removeAt(0);
// myDoubleLL.print();
// console.log(myDoubleLL.find(1));

export default DoubleLinkedList;
